progs/mcnode/main.go

```diff
@@ -177,13 +187,160 @@ func (node *Node) doLookup(ctx context.Context, pid p2p_peer.ID) (empty p2p_psto
 	return pinfo, nil
 }
 
+func (node *Node) httpPublish(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	ns := vars["namespace"]
+
+	rbody, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		log.Printf("http/publish: Error reading request body: %s", err.Error())
+		return
+	}
+
+	// just simple statements for now
+	sbody := new(pb.SimpleStatement)
+	err = json.Unmarshal(rbody, sbody)
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		fmt.Fprintf(w, "Error: %s\n", err.Error())
+		return
+	}
+
+	sid, err := node.doPublish(ns, sbody)
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		fmt.Fprintf(w, "Error: %s\n", err.Error())
+		return
+	}
+
+	fmt.Fprintln(w, sid)
+}
+
+var BadStatementBody = errors.New("Unrecognized statement body")
+
+func (node *Node) doPublish(ns string, body interface{}) (string, error) {
+	stmt := new(pb.Statement)
+	pid := node.ID.Pretty()
+	ts := time.Now().Unix()
+	counter := node.stmtCounter()
+	stmt.Id = fmt.Sprintf("%s:%d:%d", pid, ts, counter)
+	stmt.Publisher = pid // this should be the pubkey when we have ECC keys
+	stmt.Namespace = ns
+	stmt.Timestamp = ts
+	switch body := body.(type) {
+	case *pb.SimpleStatement:
+		stmt.Body = &pb.Statement_Simple{body}
+
+	case *pb.CompoundStatement:
+		stmt.Body = &pb.Statement_Compound{body}
+
+	case *pb.ArchiveStatement:
+		stmt.Body = &pb.Statement_Archive{body}
```

Comment:
Just the statement constructor that needs to wrap, so that all kind of statements can be handled by doPublish. There is no api for anything other than simple statements for now.